kikuchipy/indexation/similarity_metrics.py

```diff
@@ -0,0 +1,394 @@
+        templates: Union[np.ndarray, da.Array],
+    ):
+        nav_shape = _get_nav_shape(patterns)
+        sig_shape = _get_sig_shape(patterns)
+        nav_flat_size, sig_flat_size = np.prod(nav_shape), np.prod(sig_shape)
+        patterns = patterns.reshape((nav_flat_size, sig_flat_size))
+        templates = templates.reshape((-1, sig_flat_size))
+        return self._metric_func(patterns, templates)
+
+
+def expand_dims_to_many_to_many(p, t, flat):
+    """Expand the dims of patterns and templates to match MetricScope.MANY_TO_MANY.
+
+    Parameters
+    ----------
+    p : Union[np.ndarray, da.Array]
+        Patterns
+    t : Union[np.ndarray, da.Array]
+        Templates
+    flat : bool, optional
+        Wheter `p` and `t` are flattened
+    Returns
+    -------
+    (p,t)
+        Tuple of p and t
+        with their dimensions expanded to match MetricScope.MANY_TO_MANY.
+    """
+    metric_cls = FlatSimilarityMetric if flat else SimilarityMetric
+    p_scope_ndim, t_scope_ndim = metric_cls._SCOPE_TO_P_T_NDMI[
+        MetricScope.MANY_TO_MANY
+    ]
+    p = p[(np.newaxis,) * (p_scope_ndim - p.ndim)]
+    t = t[(np.newaxis,) * (t_scope_ndim - t.ndim)]
+    return p, t
+
+
+def zero_mean(p, t, flat=False):
+    """Subtract the mean from patterns and templates of any scope.
+
+    Parameters
+    ----------
+    p : Union[np.ndarray, da.Array]
+        Patterns
+    t : Union[np.ndarray, da.Array]
+        Templates
+    flat : bool, optional
+        Wheter `p` and `t` are flattened, by default False
+
+    Returns
+    -------
+    (p,t)
+        Tuple of p and t with their mean subtracted.
+    """
+    squeeze = 1 not in p.shape + t.shape
+    p, t = expand_dims_to_many_to_many(p, t, flat)
+    p_mean_axis = 1 if flat else (2, 3)
+    t_mean_axis = 1 if flat else (1, 2)
+    p -= p.mean(axis=p_mean_axis, keepdims=True)
+    t -= t.mean(axis=t_mean_axis, keepdims=True)
+
+    if squeeze:
+        return (
+            p.squeeze(),
+            t.squeeze(),
+        )
+    else:
+        return p, t
+
+
+def normalize(p, t, flat=False):
+    """Normalize patterns and templates of any scope.
+
+    Parameters
+    ----------
+    p : Union[np.ndarray, da.Array]
+        Patterns
+    t : Union[np.ndarray, da.Array]
+        Templates
+    flat : bool, optional
+        Wheter `p` and `t` are flattened, by default False
+
+    Returns
+    -------
+    (p,t)
+        Tuple of p and t divided respectively by their L2-norms.
+    """
+    squeeze = 1 not in p.shape + t.shape
+    p, t = expand_dims_to_many_to_many(p, t, flat)
+    p_sum_axis = 1 if flat else (2, 3)
+    t_sum_axis = 1 if flat else (1, 2)
+    p /= (p ** 2).sum(axis=p_sum_axis, keepdims=True) ** 0.5
+    t /= (t ** 2).sum(axis=t_sum_axis, keepdims=True) ** 0.5
+
+    if squeeze:
+        return (
+            p.squeeze(),
+            t.squeeze(),
+        )
+    else:
+        return (p, t)
+
+
+SIMILARITY_METRICS: Dict[str, Callable] = {}
+
+
+def _zncc_einsum(
```

Comment:
The function signature says it supports numpy arrays, would it be as simple as checking for the array types (`isinstance(patterns, np.ndarray)`) and calling either da.einsum or np.tensordot? I think we should support this.

Reply:
This now returns numpy arrays if both patterns and templates are numpy arrays.